druntime/src/core/internal/array/duplication.d

```diff
@@ -21,9 +21,9 @@ U[] _dup(T, U)(scope T[] a) pure nothrow @trusted if (__traits(isPOD, T))
     {
         import core.stdc.string : memcpy;
         import core.internal.array.construction: _d_newarrayUPureNothrow;
-        auto arr = _d_newarrayUPureNothrow!T(a.length, is(T == shared));
+        auto arr = _d_newarrayUPureNothrow!U(a.length, is(U == shared));
```

Comment:
After a quick look, I guess the 2nd `isShared` param isn't required. It's forwarded to `__setArrayAllocLength()`, but I think there it'd only matter for array `length` updates (using CAS for `isShared` then), not setting the initial length after a fresh GC allocation. But I'm not 100% sure, so being conservative here.

Reply:
Yeah, I think you are right -- `isShared` can be false always, even if the ultimate type is shared. Obviously, the block cannot be shared, as it's new!

The right place to fix that is inside `_d_newarrayUPureNothrow`. However, probably an enhancement for a future PR, not this one.

Reply:
* `_d_newarrayUPureNothrow` is a template, so it can check whether the expected return type `U` is `shared` or not, so I agree that this parameter is `isShared` redundant.
* I could see this information being useful in case we had thread-local GC heaps, in addition to the currently single global one

Anyway I agree that this is a topic for a separate PR.